### PR TITLE
[BEAM-6753] Set the stage to make schema coder update compatible

### PR DESCRIPTION
**Pull request by reuvenlax**
Adds a proto representation of schemas.
For Dataflow, embeds this proto in the SchemaCoder representation.
Adds a notion of encoding location, no longer tied to id (not yet hooked into RowCoder.encode). 
Length prefix RowCoder

### CODE REVIEW

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
@dpmills

**Comment by reuvenlax** [comment]
R: @dpmills

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java [hunk 1]
```diff
@@ -285,15 +289,24 @@ public InstrumentedType prepare(InstrumentedType instrumentedType) {
     // per-field Coders.
     static Row decodeDelegate(Schema schema, Coder[] coders, InputStream inputStream)
         throws IOException {
+      int fieldCount = VAR_INT_CODER.decode(inputStream);
```
Comment (dpmills) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java]: Can we leave this out by just using the length of the BitSet?

Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java]: Unfortunately no. BitSet.size() returns the total space used, which rounds up to the next word size.

Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java]: as discussed, we don't want to have to reorder fields in the encoding.

**Comment by dpmills** [comment]
LGTM

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
Run Portable_Python PreCommit

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 2]
```diff
@@ -635,6 +635,61 @@ message StandardCoders {
...
+  enum TypeName {
+    BYTE = 0;
+    INT16 = 1;
+    INT32 = 2;
+    INT64 = 3;
+    DECIMAL = 4;
+    FLOAT = 5;
+    DOUBLE = 6;
+    STRING = 7;
+    DATETIME = 8;
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Adding schemas to portability needs a design doc and discussion on dev@. Did I miss that?

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 3]
```diff
@@ -635,6 +635,61 @@ message StandardCoders {
...
+    BOOLEAN = 9;
+    BYTES = 10;
+    ARRAY = 11;
+    MAP = 13;
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Notably, you don't need this. `FieldType` has a oneof and one of the alternatives can be an atomic type and the others are map, row, array, logical. I would recommend breaking out more types in this oneof, for example a BytesTypes we may likely want metadata indicating a coder. And we discussed - and I think should go ahead with - removing datetime as a primitive type.

**Comment by reuvenlax** [comment]
FYI this was added right now simply for the DataflowRunner; we used the
portability protos for convenience (and because that's where it will go
eventually anyway). pushing things into portability for real indeed does
require a published design first, which is why the portability runners
currently ignore this.

**Pull request merged**